valhalla/baldr/time_info.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include <chrono>
+
+#include <valhalla/baldr/datetime.h>
+#include <valhalla/baldr/graphid.h>
+#include <valhalla/baldr/graphreader.h>
+#include <valhalla/midgard/logging.h>
+#include <valhalla/proto/api.pb.h>
+
+namespace dt = valhalla::baldr::DateTime;
+namespace sc = std::chrono;
+
+namespace valhalla {
+namespace baldr {
+// A structure for tracking time information as the route progresses
+struct TimeInfo {
+  // whether or not the provided location had valid time information or not
+  uint64_t valid : 1;
+
+  // index into the timezone database of the location
+  // used to do timezone offset as the route progresses
+  uint64_t timezone_index : 9;
+
+  // seconds from epoch adjusted for timezone at the location
+  // used to do local time offset as the route progresses
+  uint64_t local_time : 54;
+
+  // the ordinal second from the beginning of the week (starting monday at 00:00)
+  // used to look up historical traffic as the route progresses
+  uint64_t second_of_week : 20;
+
+  // the distance in seconds from now
+  uint64_t seconds_from_now : 43;
+
+  // the sign bit for seconds_from_now
+  uint64_t negative_seconds_from_now : 1;
+
+  // a timezone offset cache because doing the offset math is expensive
+  baldr::DateTime::tz_sys_info_cache_t* tz_cache;
+
+  /**
+   * Helper function to initialize the object from a location. Uses the graph to
+   * find timezone information about the edge candidates at the location. If the
+   * graph has no timezone information or the location has no edge candidates the
+   * default timezone will be used (if unspecified UTC is used). If no datetime
+   * is provided on the location or an invalid one is provided the TimeInfo will
+   * be invalid.
+   *
+   * @param location                 location for which to initialize the TimeInfo
+   * @param reader                   used to get timezone information from the graph
+   * @param default_timezone_index   used when no timezone information is available
+   * @return the initialized TimeInfo
+   */
+  static TimeInfo
+  make(valhalla::Location& location,
```

Comment:
all of the algorithms now start off by calling this function. it sets all the time information that we need to track over the course of the expansion, see the members above.